R/pre_test_power.R

```diff
@@ -1997,3 +1506,614 @@ GeoLiftPower <- function(data,
+  results$location <- lapply(results$location, function(x) paste(x, collapse = ", "))
+  results$location <- unlist(results$location)
+
+  # Step 2 - Compute Significant & Remove Duplicates
+  results <- results %>%
+    dplyr::mutate(significant = ifelse(pvalue < 0.1, 1, 0)) %>%
+    #   dplyr::filter(significant > 0) %>%
+    dplyr::distinct()
+
+  # Step 3 - Compute average Metrics
+  results <- results %>%
+    dplyr::group_by(location, duration, EffectSize) %>%
+    dplyr::summarise(
+      power = mean(significant),
+      AvgScaledL2Imbalance = mean(ScaledL2Imbalance),
+      Investment = mean(investment),
+      AvgATT = mean(att_estimator),
+      AvgDetectedLift = mean(detected_lift), .groups = "keep"
+    )
+
+  # Step 4 - Find the MDE that achieved power
+  resultsM <- NULL
+
+  for (locs in unique(results$location)) {
+    for (ts in treatment_periods) { # for(ts in treatment_periods)
+      resultsFindAux <- results %>% dplyr::filter(location == locs & duration == ts & power > 0.8)
+
+      if (min(effect_size) < 0) { # if ( min(effect_size) < 0){
+        resultsFindAux <- resultsFindAux %>% dplyr::filter(EffectSize != 0)
+        MDEAux <- suppressWarnings(max(resultsFindAux$EffectSize))
+        resultsFindAux <- resultsFindAux %>% dplyr::filter(EffectSize == MDEAux)
+      } else {
+        MDEAux <- suppressWarnings(min(resultsFindAux$EffectSize))
+        resultsFindAux <- resultsFindAux %>% dplyr::filter(EffectSize == MDEAux)
+      }
+
+      if (MDEAux != 0) { # Drop tests significant with ES = 0
+        resultsM <- resultsM %>% dplyr::bind_rows(resultsFindAux)
+      }
+    }
+  }
+
+  # Step 5 - Add Percent of Y in test markets
+  # Step 5.1 - Create the overall prop
+  AggYperLoc <- data %>%
+    dplyr::group_by(location) %>%
+    dplyr::summarize(Total_Y = sum(Y))
+
+  # Step 5.2 - Attach to Table
+  resultsM$ProportionTotal_Y <- 1
+  resultsM$Locs <- strsplit(stringr::str_replace_all(resultsM$location, ", ", ","), split = ",")
+
+  for (row in 1:nrow(resultsM)) {
+    resultsM$ProportionTotal_Y[row] <- as.numeric(AggYperLoc %>%
+      dplyr::filter(location %in% resultsM$Locs[[row]]) %>%
+      dplyr::summarize(total = sum(Total_Y))) /
+      sum(AggYperLoc$Total_Y)
+  }
+
+  # Step 6 - Remove any duplicates
+  resultsM <- resultsM %>%
+    dplyr::group_by(location, duration) %>%
+    dplyr::slice_max(order_by = power, n = 1)
+
+  # Step 7 - Create Rank variable - Adding New Ranking System
+  abs_lift_in_zero <- true_lift <- NULL # might change to es
+  resultsM$abs_lift_in_zero <- round(abs(resultsM$AvgDetectedLift - resultsM$EffectSize), 3)
+
+  resultsM <- as.data.frame(resultsM) %>%
+    dplyr::mutate(
+      rank_mde = dplyr::dense_rank(abs(EffectSize)),
+      rank_pvalue = dplyr::dense_rank(power),
+      rank_abszero = dplyr::dense_rank(abs_lift_in_zero)
+    )
+
+  resultsM$rank <- rank(
+    rowMeans(resultsM[, c("rank_mde", "rank_pvalue", "rank_abszero")]),
+    ties.method = "min"
+  )
```

Comment:
Artur, since we are making changes to the data that could affect the ranking, instead of reranking after every change, can we just rank once, at the end of all of the potential changes? Maybe line 2055 is a good place to do so.